src/vocabulary.rs

```diff
@@ -120,19 +128,62 @@ pub type Version = u32;
 pub type Datom = (Entid, Entid, TypedValue);
 
 /// `Attribute` instances not only aren't named, but don't even have entids.
+///
 /// We need two kinds of structure here: an abstract definition of a vocabulary in terms of names,
 /// and a concrete instance of a vocabulary in a particular store.
+///
+/// `Definition` is the former, and `Vocabulary` is the latter.
+///
 /// Note that, because it's possible to 'flesh out' a vocabulary with attributes without bumping
 /// its version number, we need to know the attributes that the application cares about -- it's
 /// not enough to know the name and version. Indeed, we even care about the details of each attribute,
 /// because that's how we'll detect errors.
-#[derive(Debug)]
+///
+/// `Definition` includes two additional fields: functions to run if this vocabulary is being
+/// upgraded. `pre` and `post` are run before and after the definition is transacted against the
+/// store. Each is called with the existing `Vocabulary` instance so that they can do version
+/// checks or employ more fine-grained logic.
+#[derive(Clone)]
 pub struct Definition {
     pub name: NamespacedKeyword,
     pub version: Version,
     pub attributes: Vec<(NamespacedKeyword, Attribute)>,
+    pub pre: fn(&mut InProgress, &Vocabulary) -> Result<()>,
+    pub post: fn(&mut InProgress, &Vocabulary) -> Result<()>,
+}
+
+impl Definition {
+    pub fn no_op(_ip: &mut InProgress, _from: &Vocabulary) -> Result<()> {
+        Ok(())
+    }
+
+    pub fn new<N, A>(name: N, version: Version, attributes: A) -> Definition
+    where N: Into<NamespacedKeyword>,
+          A: Into<Vec<(NamespacedKeyword, Attribute)>> {
+        Definition {
+            name: name.into(),
+            version: version,
+            attributes: attributes.into(),
+            pre: Definition::no_op,
+            post: Definition::no_op,
+        }
+    }
+
+    /// Called with an in-progress transaction and the previous vocabulary version
+    /// if the definition's version is later than that of the vocabulary in the store.
+    fn pre(&self, ip: &mut InProgress, from: &Vocabulary) -> Result<()> {
```

Comment:
In the interests of good usage documentation, perhaps we could show an example usage here and for `post`? I know we have some in the tests, but external users won't necessarily have access to those and these make more sense when seeing how they are used.